File: C-Plus-Plus/search/Jump_Search.cpp
/*JUMP SEARCH:
Jump Search is a searching algorithm for sorted arrays. 
The basic idea is to check fewer elements (than linear search) by jumping ahead by fixed steps or skipping some elements in place of searching all elements.
*/

#include<iostream>
#include<cmath>

using namespace std;

int jump_search(int a[], int n, int find)
{
    int step = sqrt(n);
    int pre_step = 0;
    while (a[min(step, n) - 1]  < find) 
    { 
        pre_step = step; 
        step += sqrt(n); 
        if (pre_step >= n)
        {
            return -1;	
        } 
    }
    while(a[pre_step] < find)
    {
        pre_step++;
        if(pre_step == min(step,n))
        {
            return -1;
        }
    }
    if(a[pre_step] == find)
    {
        return pre_step;
    }
    else
    {
        return -1;
    }
}

int main()
{
    int n,search;
    cout << "Enter no. of elements in an array: ";
    cin >> n;
    int arr[n];
    cout << "Enter elements in an array in increasing order: ";
    for(int i=0;i<n;i++)
    {
        cin >> arr[i];
    }
    cout << "Enter a number to be searched: ";
    cin >> search;
    int result = jump_search(arr,n,search);
    if(result == -1)
    {
        cout << search << " is not present in the array." << endl;
    }
    else
    {
        cout << search << " is present at index no " << result+1 << endl;
    }
    return 0;
}

/*
SAMPLE INPUT:
Enter no. of elements in an array: 
12
Enter elements in an array in increasing order: 
2 4 6 8 10 11 14 17 20 33 36 40
Enter a number to be searched: 
17

SAMPLE OUTPUT:
17 is present at index no 8
*/

File: C-Sharp/search/JumpSearch.cs
/*
C# program for jump Search
Jump Search -
Jump Search is a searching algorithm for sorted arrays. The basic idea
is to check fewer elements (than linear search) by jumping ahead by 
fixed steps or skipping some elements in place of searching all elements.
*/
using System;

public class JumpSearch
{
    public static int JumpSearchAlgo(int[] arr, int x)
    {
        int n = arr.Length;

        //Finding Block size to be jumped
        int step = (int)Math.Floor(Math.Sqrt(n));

        //Finding the block where element is present (if it is present)
        int prev = 0;
        while (arr[Math.Min(step, n) - 1] < x)
        {
            prev = step;
            step += (int)Math.Floor(Math.Sqrt(n));
            if (prev >= n)
                return -1;
        }

        //Doing a linear search for x in block beginning with prev.
        while (arr[prev] < x)
        {
            prev++;

            //if we reached next block or end of array,element is not present.
            if (prev == Math.Min(step, n))
                return -1;

        }

        //if element is found 
        if (arr[prev] == x)
            return prev;

        return -1;
    }

    public static void Main()
    {
        Console.WriteLine("Enter size of array that you like to create");
        int n = int.Parse( Console.ReadLine());
        
        int[] arr = new int[n];
        Console.WriteLine("Enter values in array");
        
        for (int i = 0; i < n; i++)
        {
            arr[i] = int.Parse( Console.ReadLine());
        }

        Console.WriteLine("Enter value that you like to find in array");
        int x = int.Parse(Console.ReadLine());

        int index = JumpSearchAlgo(arr, x);
        if (index >= 0)
            Console.WriteLine("Found at index: " + index);
        else
            Console.WriteLine(x + " isn't present in the array");
    }
}
/*
Sample Input
Enter size of array that you like to create
10
Enter values in array
12
23
34
45
56
67
78
89
90
100
Enter value that you like to find in array
45

Sample Output
Found at index: 3

Time Complexity  - O(√n) 
Space Complexity - O(1)

*/
